transport_nantes/topicblog/models.py

```diff
@@ -664,3 +664,121 @@ class TopicBlogEmailClicks(models.Model):
+            # that are draft or retired.
+            ("tbp.may_view", "May view unpublished TopicBlogPresss"),
+
+            # Granting edit permission to users does not in itself
+            # permit them to publish or retire, so it is reasonably
+            # safe.
+            ("tbp.may_edit", "May create and modify TopicBlogPresss"),
+
+            # Finally, we can grant users permission to publish, to
+            # self-publish (implies tbp_may_publish), to send, and to
+            # self-send.
+            ("tbp.may_publish", "May publish TopicBlogPresss"),
+            ("tbp.may_publish_self", "May publish own TopicBlogPresss"),
+            ("tbp.may_send", "May send TopicBlogPresss"),
+            ("tbp.may_send_self", "May send own TopicBlogPresss"),
+        )
+
+    subject = models.CharField(max_length=80, blank=True)
+    header_image = models.ImageField(
+        upload_to='header/', blank=True,
+        help_text='résolution recommandée : 1600x500')
+
+    # Content #######################################################
+    body_text_1_md = models.TextField(blank=True)
+    body_image_1 = models.ImageField(
+        upload_to='body/', blank=True,
+        help_text='résolution recommandée : 1600x500')
+    body_image_1_alt_text = models.CharField(max_length=100, blank=True)
+
+    # Plus slug, template, title, comment, and social media fields,
+    # provided through abstract base class.
+
+
+    def get_absolute_url(self):
+        """Provide a link to view this object (by slug and id).
+        """
+        if self.slug:
+            return reverse("topicblog:view_press_by_pkid",
+                           kwargs={"pkid": self.pk,
+                                   "the_slug": self.slug})
+        else:
+            return reverse("topicblog:view_press_by_pkid_only",
+                           kwargs={"pkid": self.pk})
+
+    def get_edit_url(self):
+        """Provide a link to edit this object (by slug and id).
+        """
+        if not self.slug:
+            return reverse("topicblog:edit_press_by_pkid",
+                           kwargs={"pkid": self.pk})
+        else:
+            return reverse("topicblog:edit_press",
+                           kwargs={"pkid": self.pk,
+                                   "the_slug": self.slug})
+
+
+
+class TopicBlogPressSendRecord(models.Model):
+
+    """Represent the fact that we sent an press.
+    """
+    slug = models.SlugField(max_length=90, allow_unicode=True, blank=True)
+    mailinglist = models.ForeignKey(MailingList, on_delete=models.PROTECT)
+    recipient = models.ForeignKey(User, on_delete=models.PROTECT)
+    send_time = models.DateTimeField(auto_now=True)
+    # Open time is the time of the first instance of a beacon responding.
+    open_time = models.DateTimeField()
+    # Click time is the time of the first instance of a link being clicked.
+    click_time = models.DateTimeField()
```

Comment:
When you say "of a link" you mean any link contained into the article ?

Reply:
When we provide links (URLs) in emails, we will want to automatically decorate them with something that allows us to know they've been clicked.  How we do that is for tomorrow (well, later this week), not today.

One way is to add an argument (`?something=value`).  Another way is to wrap it, `https://www.mobilitains.fr/tb/redirect/<str:value>/<str:url-encoded-destination>`.  The redirect (need better name) path reads the value and knows it refers to a given send record, and it reads the next argument and 302's to it (assuming first arg is valid and the redirect goes to us).

Reply:
The get argument sounds simpler, although fakeable, but it shouldn't be a matter of concern, would it ?

Reply:
The principle convenience of the second approach, I think, is that the first requires intercepting _all_ requests to check for the flag whilst the second only engages on specific request.

Reply:
So your advice is the make a token, embed it in email urls (so all urls will point to `https://www.mobilitains.fr/tb/redirect/<str:value>/<str:url-encoded-destination>`, reverse the token and extract the information from there.

Reply:
That's correct.

Let's check in in when you start implementing this (at the level that you've had a rough stab at it so that your head's in it) and we can make sure we're thinking the same things.